### DISCUSSION TITLE
Training Tutorial

### DISCUSSION THREAD

**Discussion by Davidmolin4** [Information-Fusion-Lab-Umass/NuX#14] [Q&A]
Hello everybody!

I would like to know if there is any tutorial about how to train normalizing flows with this library. I've been looking but I haven't found any, I couldn't understand the FlowTrainer class either.

Greetings

David

**Comment by EddieCunningham** [comment]
Thanks for the interest in NuX!  The flows are built to be as flexible as possible, so you could in theory train them using any of the training frameworks out there for JAX (optax, flax, haiku, etc.).   You'd create and initialize the flow using the format in the [readme example](https://github.com/Information-Fusion-Lab-Umass/NuX/blob/master/examples/readme_example.py) which is roughly 
``` python
# 1. Create flow
flow = nux.Sequential(...., prior=...)

# 2. Initialize the flow with data
z, log_px = flow(initialization_data, params=None, rng_key=initialization_key)

# 3. Retrieve the initialized parameters
params = flow.get_params()

# 4. Give your JAX framework the parameters
# (Depends on framework.  See [here](https://github.com/Information-Fusion-Lab-Umass/NuX/blob/master/examples/haiku_vae.py#L83) for a Haiku example)

# 5. Use the flow in your code
z, log_px = flow(x, params=params, rng_key=rng_key)
```


The FlowTrainer class is another way to train a model.  You can take a look at [this](https://github.com/Information-Fusion-Lab-Umass/NuX/blob/master/examples/celeba_ipcf.py) for an example of how to use it, but to be honest I think training with some of the larger frameworks might be the best way to go because they are more developed.  However if you're still interested in how to train flows using the class in this library and want to learn more, I'd be happy to explain more about it.

**Reply by Davidmolin4** [reply]
Thank you for your response. I will try with that

Regards

**Reply by Davidmolin4** [reply]
I have the following problem: When training a RationalQuadraticSpline() as a flow, the loss function takes the value nan. This does not happen with other type of flows (LogisticCDFMixtureLogit() for example).

The code I made is the following
```python
import jax
import jax.numpy as jnp
import nux

#Artificial data
rng_key = jax.random.PRNGKey(0)
batch_size, dim = 10000, 1
x = jax.random.normal(rng_key, (batch_size, dim))

# Initialize the flow with data-dependent initialization
flow = nux.Sequential(nux.RationalQuadraticSpline(),
prior = nux.UnitGaussianPrior())
z, log_px = flow(x, rng_key=rng_key)
params = flow.get_params()

# Optimizer
import optax
optimizer = optax.adam(0.01)
opt_state = optimizer.init(params)
epochs = 2000

# Loss function
def loss(params, x, rng_key):
  z, log_px = flow(x, params=params, rng_key=rng_key)
  return -log_px.mean()

# Training
from tqdm.auto import tqdm # Bar progress
for step in tqdm(range(epochs)):
  if step%100 == 0:
    print(step,'%.2f' % (loss(params,x, rng_key)))
  grads = jax.grad(loss)(params, x, rng_key)
  updates, opt_state = optimizer.update(grads, opt_state)
  params = optax.apply_updates(params, updates)
```

**Reply by EddieCunningham** [reply]
Hmm, I'm not sure what the issue could be.   When I run your code the loss goes down to 1.43 after 300 steps and stays there.  Do you have the latest version of NuX?  Older versions have had bugs in the spline class.

Also, you can significantly speed up this example by jitting the gradient function!  You would do this by changing your training loop to
``` python
jitted_grad = jax.jit(jax.grad(loss))
for step in tqdm(range(epochs)):
  if step%100 == 0:
    print(step,'%.2f' % (loss(params,x, rng_key)))
  grads = jitted_grad(params, x, rng_key)
  updates, opt_state = optimizer.update(grads, opt_state)
  params = optax.apply_updates(params, updates)
```

**Reply by Davidmolin4** [reply]
Thanks again Eddie

I installed nux in colab through the following command

```python
!git clone --recursive https://github.com/Information-Fusion-Lab-Umass/NuX.git
```

But I can't tell which version has been installed

**Reply by EddieCunningham** [reply]
Can you share a copy of the colab notebook?  I'm not able to reproduce the error.

**Reply by Davidmolin4** [reply]
I discovered the error.

I was using the following data.

```python
# No gaussian data
rng_key = jax.random.PRNGKey(0)
batch_size, dim = 10000, 1
x = jax.random.normal(rng_key, (batch_size, dim))
x = jnp.concatenate((x, jax.random.normal(rng_key, (batch_size, dim))+10))
```

With the default parameter bounds for the spline. Changing the bounds according the data solves the issue.

I wrote you in the question another data. My mistake.

**Reply by EddieCunningham** [reply]
No problem, I'm glad you found the issue
